### ISSUE TITLE
Change the domain and range of isVariantOf to Product

### ISSUE THREAD

**Issue by jvandriel** [schemaorg/schemaorg#1797]
(edit by @danbri June 2020; see also: #2587)

One of the things I keep running into over and over again is that there currently is no way to express a _'variant'_ relation between products without the use of ```ProductModel```. Which IMO is an serious issue as there are far more products out there of which publishers don't have info about / a resource for the ```ProductModel``` than there are publishers that do have such info.

For example, cell phone cases - a type of product that often is available in several colors yet for which, more often than not, doesn't exist a resource containing product model information, eg:
- https://www.tech21.com/en_us/evo-check-s8-rose-tint-white
- https://www.tech21.com/en_us/evo-check-s8-light-blue-white
- https://www.tech21.com/en_us/evo-check-s8-clear-white
- https://www.tech21.com/en_us/evo-check-s8-smokey-black

4 variants (color) of the same product of which the ```ProductModel.name``` is 'Evo Check' according to their own structured data, but that's all the info there is. 

Currently such products can only be sort-of chained together by using either the ```isSimilarTo``` or ```isRelatedTo``` properties, which IMHO, don't express the correct relation (and ```sameAs``` doesn't do it for me either).

And so I suggest we change the domain and range of ```isVariantOf``` from ```ProductModel``` to ```Product```.

Interesting note:
This topic became top of mind for me again when I was studying Google's [Merchant Center documentation](https://support.google.com/merchants/answer/6386198), which contains an example of an array of unchained products that implies a group of product variants on a single page (try adding a ```mainEntityOfPage``` to that example, which is important if there are also other top-level entities on a page, besides the product variants). 

So the way I see it is that even Google isn't sure how to handle the relation between product variants and uses an array as a poor man's solution to come to some sort of _implied_ semantics where _explicit_ semantics are needed.

**Comment by jvandriel** [comment]
What I'm hoping to achieve is that when ```ProductModel```info isn't available publishers still are able to specify product variant as such in a:

*Multi page environment*
```
{
  "@context":"http://schema.org",
  "@type":"Product",
  "name":"Evo Check Case for Galaxy S8 - Rose Tint/White",
  "url":"https://www.tech21.com/en_us/evo-check-s8-rose-tint-white",
  "isVariantOf":
  [
    {
      "@type":"Product",
      "url":"https://www.tech21.com/en_us/evo-check-s8-light-blue-white"
    },
    {
      "@type":"Product",
      "url":"https://www.tech21.com/en_us/evo-check-s8-clear-white"
    },
    {
      "@type":"Product",
      "url":"https://www.tech21.com/en_us/evo-check-s8-smokey-black"
    }
  ]
}
```

*Single page environment*
```
{
  "@context":"http://schema.org",
  "@type":"Product",
  "name":"Evo Check Case for Galaxy S8",
  "color":"Rose Tint/White",
  "sku":"T21-5664",
  "url":"https://www.tech21.com/en_us/evo-check-s8",
  "isVariantOf":
  [
    {
      "@type":"Product",
      "color":"Light Blue/White",
      "sku":"T21-5586"
    },
    {
      "@type":"Product",
      "color":"Clear/White",
      "sku":"T21-5584"
    },
    {
      "@type":"Product",
      "name":"Smokey/Black",
      "sku":"T21-5585"
    }
  ]
}
```

**Comment by jvandriel** [comment]
And if we want to be helpful in regards to ease of implementation for publishers than we help them avoid the needed use for fragment identifiers, as shown in the above examples, than we also introduce an inverse property for ```isVariantOf```, namely: ```hasVariant```.

With a description of:
> A pointer to a product variant from which this product is the base product. It is safe to infer that the variant inherits all product features from the base model, unless defined locally. This is not transitive.

Making the following possible for a:
*Multi page environment*
```
{
  "@context":"http://schema.org",
  "@type":"Product",
  "name":"Evo Check Case for Galaxy S8 - Rose Tint/White",
  "color":"Rose Tint/White",
  "sku":"T21-5664",
  "url":"https://www.tech21.com/en_us/evo-check-s8-rose-tint-white",
  "hasVariant":
  [
    {
      "@type":"Product",
      "url":"https://www.tech21.com/en_us/evo-check-s8-light-blue-white"
    },
    {
      "@type":"Product",
      "url":"https://www.tech21.com/en_us/evo-check-s8-clear-white"
    },
    {
      "@type":"Product",
      "url":"https://www.tech21.com/en_us/evo-check-s8-smokey-black"
    }
  ]
}
```

*Single page environment*
```
{
  "@context":"http://schema.org",
  "@type":"Product",
  "name":"Evo Check Case for Galaxy S8",
  "color":"Rose Tint/White",
  "sku":"T21-5664",
  "url":"https://www.tech21.com/en_us/evo-check-s8",
  "hasVariant":
  [
    {
      "@type":"Product",
      "color":"Light Blue/White",
      "sku":"T21-5586"
    },
    {
      "@type":"Product",
      "color":"Clear/White",
      "sku":"T21-5584"
    },
    {
      "@type":"Product",
      "name":"Smokey/Black",
      "sku":"T21-5585"
    }
  ]
}
```

**Comment by jvandriel** [comment]
- _Done tinkering with the code examples - free to shoot on 'm now_

**Comment by vberkel** [comment]
I would like to propose a different way of solving this issue. Could we swap the rdfs:subClassOf hierarchy of Product and ProductModel?  

As we have it today, the ProductModel is described as **prototypical**, as in something that is more abstract than the Product itself:
> "A datasheet or vendor specification of a product (in the sense of a **prototypical** description)".  

Yet, today's hierarchy goes from most abstract to concrete then back to abstract.
`Thing > Product > ProductModel `

In concrete terms, we can say today that all ProductModel data items are also implied to be a Product and a Thing. A Product conversely cannot be inferred to be a ProductModel even though all products are based on some conceptualization. I don't think it is correct that a prototype, abstract, template-able sort of thing is also through inheritance a kind of concrete thing. 

If we flipped the hierarchy:
`Thing > ProductModel > Product`
We go from most abstract to concrete. A concrete, sellable product is also implied to based on some ProductModel (and Thing). 

If we agree to the above, we might want to discuss which properties should be reassigned to a domain of ProductModel from Product. I would start with the assumption that all properties of Product get reassigned to ProductModel. This would solve @jvandriel issue of changing isVariantOf through inheritance. 

I'm curious what @mfhepp would have to say about this.

**Comment by thadguidry** [comment]
@vberkel Its not ProductORModel ... its actually AProductModel
So the heirarchy is fine the way it is.
What needs to change is our current stupid confusing description. Which I never did agreed to. :)

The better description and what I alias in my apps is:
"A particular design or version of a Product."

**Comment by jvandriel** [comment]
Just to make sure I understand you correctly @thadguidry...

```ProductModel``` = Product variant?

Cause if so, that's not how I understand ```ProductModel```, and I'm pretty sure most publishers don't get it that way either, eg. the example I showed above refers to it as if it's a prototypical/abstract and nearly all sites I checked on their use of ```ProductModel``` do so.

And if ```ProductModel``` = Product variant, than it might be an idea somebody talks to the major search engines as they sure don't treat it as if it's a _"A particular design or version of a Product."_ given the fact none of 'm generate rich snippets for it (which only makes sense to me if it's an abstract).

**Comment by thadguidry** [comment]
@jvandriel ProductModel = model (which could be a variant of another model)

For clarity, scroll to the bottom and see the reversing properties
http://schema.org/ProductModel

In Good Relations and Schema.org...
Products can have Models
Models can have variants. (another Model)

- Product = iPhone
  - ProductModel = iPhone 5
  - ProductModel = iPhone 6
  - ProductModel = iPhone 7
  - ProductModel = iPhone 6 S
    - isVariantOf = iPhone 6 

- IndividualProduct = iPhone 6 S
  - serialNumber = "IMEA # AA-BBBBBB-CCCCCC-D"

**Comment by jvandriel** [comment]
> In Good Relations and Schema.org...
> Products can have Models.
> Models can have variants. (another Model)

Great, at least we're talking about the same thing. But for all intends and purposes, what I'm missing in this list is that products can have variants as well, eg:

- ProductModel = iPhone 6 S
-- @id = #ProductModel
-- isVariantOf = iPhone 6
-- predecessorOf = iPhone 7
-- successorOf = iPhone 5

- Product = iPhone 6s 4.7-inch 128 GB Silver
-- model = #ProductModel
-- color = silver
-- memory = 128 GB
-- hasVariant = ???
-- isVariantOf = ???

- Product = iPhone 6s 4.7-inch 32GB Gold
-- model = #ProductModel
-- color = gold
-- memory = 32 GB
-- hasVariant = ???
-- isVariantOf = ???

Something I feel ```isSimilarTo``` & ```isRelated``` don't cover as these can refer to completely different products (of a different brand) with similar properties as well - reason why I'd like to see the domain and range of ```isVariantOf``` be changed to ```Product``` as well as adding a reverse property (```hasVariant```).

> Some will argue that iPhone is ONLY a brand... well yeah, it "can be" a brand, but its not necessary to say ONLY. Its certainly allowable to say that iPhone can be a Product, and nothing in Good Relations or Schema.org prevents that. :) Thank goodness.

Don't know if I agree with you on this one though. Reason being that it's not very often a ```ProductModel``` is up for sale - eg, Apple isn't very likely to sell it's iPhone 8 ```ProductModel``` though they do sell products. Heck, even they annotate their iPhone 8 as a ```Product``` as opposed to a  ```ProductModel```: http://bit.ly/2BhYaBO

What's the rationale behind: _" Its certainly allowable to say that iPhone can be a Product"_? (after all, one buys a certain configuration of a product and not the ```ProductModel``` itself)

And at the same time, if a ```ProductModel``` could be considered a ```Product``` than why isn't this reflected by the major search engines by allowing those to generate 'rich snippets'* as well? To me this is an indication they see a substantial difference between the two as well 

_* no, I don't want to get into a 'rich snippet' discussion over this, the example is just illustrative_

**Comment by jvandriel** [comment]
The answer to the question: > _"How to avoid repeating values that are the same for the differently sized products?"_ on [StackExchange](https://webmasters.stackexchange.com/questions/85877/schema-org-product-same-product-different-sizes-and-prices#answer-85878) illustrates perfectly one of the issues I hope to resolve with this proposal - it would eliminate the need for repeating the same properties over and over again.

Even @mfhepp's publication about [markup for product variants with different prices](https://blog.heppresearch.com/2015/07/23/schema-org-product-variants-with-different-prices/) doesn't take this into account.

**Comment by jaw111** [comment]
For the use cases I'm commonly dealing with (electronic components) I'd propose to tighten the definition of `isVariantOf` based around idea of [form, fit and function](https://en.wikipedia.org/wiki/Form,_fit_and_function), such that the (manufacturing) variants of a product model **must** have the same form, fit and function.

In that view the iPhone 6s is not a variant of the iPhone 6, but a separate model (perhaps in the same series or family).

However I can imagine consumer goods has it's own world view, same for other domains.

**Comment by danbri** [comment]
hi folks! As alluded to in #2587, @alex-jansen @rvguha and I have been discussing this with Google colleagues too. We think there is scope to clarify how to treat variants.- In line with @mfhepp in his old blog post (which seems offline but [via archive.org](https://web.archive.org/web/20160312112025/https://blog.heppresearch.com/2015/07/23/schema-org-product-variants-with-different-prices/)) we would agree that the "abstract model of the product" (omitting /size, /color etc concrete details) is worth describing separately as a distinct entity. We think this can be done more explicitly with some modest tweaks, including a new (sub)type.

- Although the idea is similar to /ProductModel data sheets, the "product model" construct addresses a slightly different situation (c.f. @jvandriel's point above, "there are far more products out there of which publishers don't have info about / a resource for the /ProductModel than there are publishers that do have such info.").
- noting that @jvandriel and others have advocated to extend /isVariantOf so that it doesn't just relate /ProductModels to /ProductModels, but also relates specific product variants to general products that stand for a group of predictably-varying products.
- and that @mfhepp's example markup instead uses an indirection through an Offer to implicitly link a general product to a specific one, i.e. a /Product ("The Hepp Smart Watch is a unique wristwatch.") would have /offers, one for each variant offered, such as "Hepp Smart Watch in Silver with 16 GB RAM", "Hepp Smart Watch in Gold with 64 GB RAM". This is fine when you have such offers to publish, but there seems be a shared desire to model the underlying product situation independently.
- we see value in introducing a new type, analogous to /ProductModel, to more explicitly represent a group or cluster of varying products. We suggest /ProductCluster for this (or /ProductGroup). It could be used in the styles advocated by either @jvandriel or @mfhepp. This would make it easier for consuming applications to understand when a Product-description was intended as a general vs specific variant, rather than leaving it ambiguous (since sparse descriptions of the variants are otherwise indistinguishable from the general descriptions).
- we also propose using /ProductCluster to describe common characteristics of the specific product variants in that group or cluster, as well as the nature of the variations (i.e. color, pattern etc.). We expect to propose a mechanism similar to that used in /StatisticalPopulation and /ProductModel for this, where properties applicable to the varying instances can be ascribed to the prototypical "item representing the whole group". 
- Suggest rewording definition of /isVariantOf to something along these lines, with the details moved out into a dedicated doc: "Indicates a base or parent product that this product is a variant of. Properties of the referenced product are generally also applicable to the variant products, i.e. the convention is to describe a ProductModel or ProductCluster as the shared prototype of all its variants. See [doc] for details."There are reasonable reasons to debate whether extending /isVariantOf is better than a new property, or for that matter whether there's a case to have /ProductModel and the proposed /ProductGroup be a sibling type to /Product rather than a subtype.  Before we spin off into discussion of those details it would be good to re-affirm the value of doing something in this broad direction.

**Comment by jvandriel** [comment]
Sounds good to me. Personally I'd prefer the name ```ProductGroup``` as it's closest to what it's called in other namespaces/specs for marketplaces, which often speak about a 'group' of products.

Any thoughts about the possibility of adding a ```hasVariant``` property so there's no dependency on the existence of offers or the need for referencing by fragment identifiers?

**Comment by danbri** [comment]
Having an inverse of /isVariantOf sounds useful to me. As for names, the terminology of "parent" also seems popular in e-commerce circles, but we have already used it in schema.org metaphorically in other places.

**Comment by thadguidry** [comment]
This is valuable.
Agree with /ProductGroup name.

Will /ProductGroup have properties for parent/sibling relationships like
/ProductModel has with pre/succ ?
Do we expand pre/succ to also expect a type of /ProductGroup in addition to
the expected /ProductModel?

I am worried about mixing Types and Multi-Typed entities playing within all
this...so...
It might be better to have new properties for the parent/sibling
relationships just for /ProductGroup to alleviate ambiguity with
/ProductModel at the predicate level?
I.E. /parentProductGroup
      /childProductGroup

Thad
https://www.linkedin.com/in/thadguidry/

**Comment by jaw111** [comment]
To me /ProductGroup is still quite vague and perhaps might just be a arbitrary group of products that share some similar feature or characteristic. How does it differ from a /category of products?

Perhaps something like /ProductLine, /ProductSeries or /ProductFamily might fit better?

Without wanting to muddy the waters further, isn't it the case that an /Offer is (logically) related to either an /IndividualProduct or /SomeProducts i.e. you (generally) do not offer the /ProductModel for sale. Isn't the use of the more generic /Product class just ambiguous in those cases?

**Comment by danbri** [comment]
@jaw111 /ProductGroup (or cluster or whatever) would be products which differed only along explicitly declared lines, e.g. w.r.t. sizes, colors.

Line/Series/Family seem closer to the "model" construct. 

The observation that the abstract product model, or product group, isn't itself offered for sale is one of the points made in favour of having them be under /Intangible. At which point it becomes a technical debt / workflow etc management issue to keep the type/property associations properly in sync. The original Good Relations design had services and products in the same type, for example; in Schema.org we ended up treating them separately, and so have to maintain by hand various places in which either could appear. It's one of the costs of trying to avoid an over-deep type hierarchy with lots of abstractions.

As to @thadguidry 's point, yes - interactions with multi-typed entities could complicate things. I was talking recently with @alex-jansen also on how this would interact with subtypes of Product, if we (hypothetically) were to go in that direction and distinguish eg WearableProduct from others.

**Comment by jaw111** [comment]
@danbri the naming of the classes seems something that'll be hard to find something to please all parties. I struggle a bit with when something would be more of a /ProductGroup or a /ProductModel. In most cases I deal with the Series is a group of products that differ along explicitly declared lines, whilst also appearing in the same data sheet.

Would it help to have some concrete uses cases to clarify when it would be more appropriate to use which and how that might be interpreted by an application that consumes that data?

**Comment by alex-jansen** [comment]
@jaw111 One key distinguishing characteristic of a /ProductGroup could be that it is **always** virtual (non-tangible). Which means that 1 or more defined attributes (the canonical examples of which are /size and /color for apparel) are needed to turn it into a real tangible product (aka a "variant"). A /ProductGroup will therefore not have the typical identifiers such as a /sku or a /gtin that a tangible item has. 

In this [example](https://www.zalando.co.uk/soyaconcept-alexa-parka-dark-army-so821u00r-n11.html) size and color are the variant attributes, while for example material and product details are common.

A /ProductModel on the contrary appears to represent a "complete" and tangible item in a series of related items (see also the comment of @thadguidry). 

The main reason to make the concept of /ProductGroup explicit was nicely phrased by @jvdriel, namely to introduce an explicit and unambiguous way of representing and grouping product variants and their variant attributes. This would make it easier for consuming applications since those would not have to reconstruct (or guess) variant groupings.

**Comment by danbri** [comment]
Some notes towards draft schemas and examples (join #1797 and #2587), although not addressing all of the goals in the other issue around SizeDetails:

https://docs.google.com/document/d/19DpcLStce-n1iF1rsIyS3Pz9_h-xSaQ1J_Nu7m24Wjo/edit#

**Comment by jaw111** [comment]
@alex-jansen ok, so using an example at hand for microchips (B2B), we might have the series [74ABT125](https://www.nexperia.com/products/analog-logic-ics/asynchronous-interface-logic/buffers-inverters-drivers/series/74ABT125.html) which would be a /ProductGroup containing the /ProductModel instances (with some variant hierarchy):

* [74ABT125BQ](https://www.nexperia.com/product/74ABT125BQ)
    * 74ABT125BQ,115
* [74ABT125D](https://www.nexperia.com/product/74ABT125D)
    * 74ABT125D,623
    * 74ABT125D,602
* [74ABT125DB](https://www.nexperia.com/product/74ABT125DB)
    * 74ABT125DB,118
* [74ABT125PW](https://www.nexperia.com/product/74ABT125PW)
    * 74ABT125PW,118
    * 74ABT125PW,112

Where the lowest level represent different packing methods for 'the same' product, like you might get 'the same' 330ml can of coke as a single item, or a 6-pack or 24-pack. As we're still not describing a specific box of chips, or 'some' anonymous boxes of chips, nor offering something for sale, then it's still appropriate to us /ProductModel for that.

Example:

```turtle
@prefix schema: <http://schema.org/>

<74ABT125> a schema:ProductGroup ;
  schema:name "74ABT125" ;
  schema: description "Quad buffer; 3-state" ;
  schema:hasVariants <74ABT125BQ>, <74ABT125D>, <74ABT125DB>,  <74ABT125PW> .

<74ABT125PW> a schema:ProductModel ;
  schema:name "74ABT125PW" ;
  schema:description "Quad buffer; 3-state in SOT402-1" ;
  schema:hasVariants <74ABT125PW,118>, <74ABT125PW,112> .

<74ABT125PW,118> a schema:ProductModel ;
  schema:name "74ABT125PW,118";
  schema:mpn "74ABT125PW,118" ;
  schema:description "Quad buffer; 3-state in SOT402-1 on Reel 13" Q1/T1 2500 pieces"
  schema:isVariantOf <74ABT125PW> .

<74ABT125PW,112> a schema:ProductModel ;
  schema:name "74ABT125PW,112";
  schema:mpn "74ABT125PW,112" ;
  schema:description "Quad buffer; 3-state in SOT402-1 Bulk pack 2400 pieces"
  schema:isVariantOf <74ABT125PW> .
```

Does that make sense?

**Comment by alex-jansen** [comment]
@jaw111 How about this to differentiate /ProductGroup and /ProductModel:

A /ProductGroup (1) never has a SKU and (2) contains a set of Products that vary on explicitly defined specific dimensions only (so it defines both the set of variants and which values must be added by those variants).

A /ProductModel (1) might have a SKU and (2) is more flexible in terms of the variations between related ProductModels (even if we disallow overruling inherited values).

A /ProductGroup is thus more restricted compared to a /ProductModel. 

For example, in the world of eCommerce, the quantity is often an identifying dimension of a variant . One can buy the same roll of Scotch Tape in 3-packs or 6-packs. Each of the packs has a separate SKU, the /ProductGroup would be "Scotch Tape" and the /Products would be the 3-pack and the 6-pack.

**Comment by danbri** [comment]
For this issue and #2587 -

* https://webschemas.org/ProductGroup and nearby,
* https://webschemas.org/pattern
* https://webschemas.org/size

This is a work in progress, but I expect to include it in the 9.0 release as a Pending change.

**Comment by thadguidry** [comment]
@danbri small quibble, but... could we get the description on https://webschemas.org/ProductGroup to also mention "features" perhaps like:

> A ProductGroup represents a group of Products that vary only in certain well-described ways, such as by, size, color, material etc., and sometimes optionally by **features**.

```
{
 "ROTO-GRIND",
 "DUAL-SHAFT",
 "QUAD-SHAFT",
}
```
1 example use case:
![image](https://user-images.githubusercontent.com/986438/85323384-8603b600-b48d-11ea-98d6-353f23b93f86.png)

**Comment by danbri** [comment]
@thadguidry what would "features" look like in terms of schema markup here?

**Comment by thadguidry** [comment]
@danbri There's a lot, depending on the ProductGroup and industrial manufacturer.  I only gave this one example to illustrate a real world scenario with 2 ProductGroup levels (actually 3 if you count the below _drive_ option which some publishers would do and how their catalogs are realized).  Anyways, the next ProductGroup feature would be 2 such options:
```
{
  "Electric drive",
  "Hydraulic drive"
}
```
and then in ProductModel you have those further feature options of the _drive horsepower_:
```
{
  "100hp",
  "125hp",
  etc.
}
```
and other Model features.

So, many times in **industrial** manufacturing, as some folks are saying above, like in silicon chip fab, the ProductGroups are sometimes heavily nested with Grouping levels by **features**, before even reaching the ProductModels that also have **features** and then finally IndividualProduct's.

I think as a good doc clue to publishers, we should inform them somehow more that they can do **feature** grouping by ProductGroup or ProductModel, depending on aggregate features in a Product series (line) and that Schema.org can handle that for them.  Right now, the descriptions and docs are not really _telling_ of that.

Does that help understand more?

**Comment by jaw111** [comment]
@alex-jansen in the example:

> For example, in the world of eCommerce, the quantity is often an identifying dimension of a variant. One can buy the same roll of Scotch Tape in 3-packs or 6-packs. Each of the packs has a separate SKU, the /ProductGroup would be "Scotch Tape" and the /Products would be the 3-pack and the 6-pack.

Then is it fair to say the 3-pack and 6-pack would each be a /ProductModel as they are not an /IndividualProduct no /SomeProducts (at least until they are put into a basket or picked off the shelf).

To put it another way, is there any case where one should ever use /Product as it should always be possible to be more specific and use the appropriate subclass instead, otherwise just make everything a /Thing and leave it at that (just to take things to the extreme).

**Comment by thadguidry** [comment]
@alex-jansen "Packages" often do have a SKU (and in the USA, a single UPC label) and treated as a sellable item (fit for sale). A Package offered **can** be an IndividualProduct, if it has a serial number (for example, to allow a user to register the individual product to their individual self)

3-pack item and 6-pack item are items being offered, so they are a /Product (fit for sale) or a subtype of /Product.

Also, Schema.org doesn't have much to offer for the supply chain side of things, since it covers only some sides of Trade and Commerce.

**Comment by jaw111** [comment]
I see that [schema:ProductGroup](https://schema.org/ProductGroup) is online, so I'm trying it out based on earlier suggestions.

A couple of possible inconsistencies that arise:

1. It's stated above that /ProductGroup would never have a /sku whilst being a subclass of /Product would imply that it may have a /sku
2. Given that /ProductGroup is a subclass of /Product then would it be appropriate to relate a /ProductGroup to another /ProductGroup using /isVariantOf or /hasVariant

**Comment by danbri** [comment]
@jaw111

on (1.) - the type hierarchy suggests that /sku might be applicable to all and any Products. But in this case, it isn't. This is similar to openingHours applying to places, and Country being a Place; or faxNumber applying to Volcano. The vocabulary structure only sets out the possibilities, it does not guarantee that there is a sensible value for a property in every instance of every type, even when the property is in general reasonable for that type.

In this case we looked at moving /ProductGroup out to sit in parallel to /Product but the benefits seemed minor enough that we didn't take that direction.

(2.) I'm not sure, I believe this possibility came up. @alex-jansen - ping

**Comment by alex-jansen** [comment]
Re.2: Yes, allowing a multi-level hierarchy as discussed earlier in this thread is explicitly supported, so a ProductGroup can have variants that are ProductGroups themselves.

**Comment by alex-jansen** [comment]
@jvandriel fair question and just a url should indeed be sufficient. We were considering adding [URL](https://schema.org/URL) to the range of [variesBy](https://schema.org/variesBy). An even better alternative could be [Property](https://schema.org/Property). Thoughts?

**Comment by jvandriel** [comment]
Well, I'm right in the middle of 3 experimental projects involving `ProductGroup` and `ProductCollection` (hoping to share my experiences once the projects are live and done, so to help with the discussions). 

One of which a is a basic webshop, for which I had to stay quite pragmatic so to prevent them from losing their Rich Results in Google search (which currently still happens when a page contains multiple products which are eligible for Rich Results).

Though the other 2 involve SaaS products with subscriptions. Which is something we never really spoke about during the discussions around the new types (we did some stabbing at subscriptions), yet for which I now discovered these new types actually make a wonderful fit. Interestingly enough I can already say that `variesBy` becomes a whole new playing field than we originally imagined it to be though. 

In 1 of the 2 cases the variances lie with different subscriptions offering different combinations of sub-applications that make up a software suite. The main type being a whopping: `["ProductGroup","ProductCollection","SoftwareApplication"]`)  of which the variances are defined by:
```
"variesBy": {
  "@type": "DefinedTerm",
  "@id": "https://schema.org/includesObject",
  "name": "includesObject",
  "url": "https://schema.org/includesObject",
  "inDefinedTermSet": "https://schema.org"
},
```
In this case though, `includesObject` seems reasonable for `ProductCollection`, which I can imagine isn't unheard of on GMC's side of things either (baskets with perfume, soap, oil, etc) and I don't expectthis to be much of an issue for any consumer of such data.

In the other case the I'm describing eSim Data Plans (`["ProductGroup","SoftwareApplication"]`). Which means the variances are actually defined within the `Offer` and `UnitPriceSpecification` properties:
```
"variesBy": [{
  "@type": "DefinedTerm",
  "@id": "https://schema.org/price",
  "name": "price",
  "url": "https://schema.org/price",
  "inDefinedTermSet": "https://schema.org"
}, {
  "@type": "DefinedTerm",
  "@id": "https://schema.org/billingIncrement",
  "name": "billingIncrement",
  "url": "https://schema.org/billingIncrement",
  "inDefinedTermSet": "https://schema.org"
 }, {
  "@type": "DefinedTerm",
  "@id": "https://schema.org/billingDuration",
  "name": "billingDuration",
  "url": "https://schema.org/billingDuration",
  "inDefinedTermSet": "https://schema.org"
}, {
  "@type": "DefinedTerm",
  "@id": "https://schema.org/referenceQuantity",
  "name": "referenceQuantity",
  "url": "https://schema.org/referenceQuantity",
  "inDefinedTermSet": "https://schema.org"
}],
```

A scenario we definitely never discussed as in the discussions thus far we've mostly been talking about variances which are a direct derivative of a `Product`'s properties; And not properties which can be found 2 or 3 triple statements deep. 

Now the above markup illustrates I am able to express this with the current setup though I have my doubts about how 'consumable' this really is (not that I have any better idea at the moment, still elbows deep in writing markup, more reflections later).

All this to say, I'm not sure - taking the last example in mind - what to do with situations where the properties describing the variances are found at a deeper level of nesting, which as I now discovered isn't all that uncommon when expressing SaaS products. Thoughts?

As for your question about adding `Property`, do you mean the actual class, including its RDFs-ish properties (which I'd be quite scared of introducing to the rest of the world, especially marketers and SEOs) or do mean to say that people would be allowed to simply provide a schema.org property label/name as value (which seams quite reasonable to me)?

**Comment by Prlme1** [comment]
Hello Dear experts,

Thank you very much for your insights. I am a start up online shop and really struggle to accomplish google requirements. Would you mind please helping?

The advocates for "hasVariantOf" does not comply with the google demands, since Google insists that each variant must have a ProductID that refers to the original ProductID. please read this google link: https://support.google.com/merchants/answer/6324507?hl=en&ref_topic=6324338#zippy=%2Cgeneral-examples

here is there citation: "Submit a unique ID for each different product. Use the same value for item group ID [item_group_id] attribute for all variants of the same product. For example, a T-shirt comes in 9 variants: 3 sizes (small, medium, large) and 3 colors (red, blue, green). Submit each variant as a separate product (for a total of 9 separate products), and submit the same value for the item group ID attribute for each to indicate that these are variants of the same product."

In addition to this, there is no [item_group_id] attribute on schema.org, so what are they referring to?

**Comment by jvandriel** [comment]
> _"there is no [item_group_id] attribute on schema.org, so what are they referring to?"_

There is, though the properties used for this are called differently:
`Product > inProductGroupWithID > Text` and `ProductGroup > productGroupID > Text`

**Comment by Prlme1** [comment]
> > _"there is no [item_group_id] attribute on schema.org, so what are they referring to?"_
> 
> There is, though the properties used for this are called differently: `Product > inProductGroupWithID > Text` and `ProductGroup > productGroupID > Text`

Dear Jvandriel, do you know whether I am allowed to use separate schema code for each of the product variant, Or I must use a single code? For example if I have 3 colors and 3 sizes this is in total 9 products, hence 9 different schemas. I am failing to put everything in a single code.

**Comment by Prlme1** [comment]
dear jvandriel,

thank you very much for your comprehensive response, while I am testing the code, would you mind helping with answering a very simply question:

what if I just ignore the complicated Schema and simply override everything in the Google Merchant Center manually, Since they have the manual option for that which I just noticed. 

Do you believe it will work?

**Comment by jvandriel** [comment]
Part of what we're doing here is trying to create an alternative for having to have XML feeds. Meaning you can either use structured data OR GMC XML feeds.

The most luxurious option for consumers of such data is if you do both. In which's case you should try to make sure both methods end up offering exactly the same info. I wouldn't count on either one of them being able to overrule the other.

**Comment by Prlme1** [comment]
dear jvandriel,

the problem that I already have the automatic crawling feed which I tried to solve the same way as you - creating the Schema that answers all the questions. But I am failing to do it. It is too complicated for me.

I still have the automatic crawling feed and try to override everything manually. 

However, I faced with another problem. Because the feed is automatic and the changes are manual I am failing to add the sizes. Because google insists for every size and color to have a separate product. But I cannot do it since its automatic feed. so shall I delete the automatic feed and do everything manually?

**Comment by jvandriel** [comment]
I'm sorry @Prlme1 but I fear I can't help you out with that as I'm not qualified to make any such statements , nor is this the right platform to be discussing that. Here we try to work on the development of schema.org itself.

If you've got any questions about how to resolve your situation I suggest you ask them over at the [GMC's help center](https://support.google.com/merchants).

**Comment by Prlme1** [comment]
thank you very much for you help. I will keep following your Schema inventions.

**Comment by Prlme1** [comment]
dear jvandriel,

I understood my mistake confusing the Schema.org and the Structural Data with the internal schema of google within the Merchant Center. I promise to stick with the topic discussed here. Would you mind helping with another confusion please about the schema:

How to make the Structural Data in a such way so that Google Merchant Center reads off each variant as an individual product. Lets say I have 3 colors and 3 sizes which is in total 9 products. Shall I just add 9 different schemas to the web page so next time google crawl it sees 9 products and adds them all to the Merchant Center? Or it should be a single code with let's say VariantOf attributes? will it be able to read off a single code as a 9 different products and adding all 9 to the merchant center?

Recall the google requirement: 

https://support.google.com/merchants/answer/6324507?hl=en&ref_topic=6324338#zippy=%2Cgeneral-examples

here is there citation: "Submit a unique ID for each different product. Use the same value for item group ID [item_group_id] attribute for all variants of the same product. For example, a T-shirt comes in 9 variants: 3 sizes (small, medium, large) and 3 colors (red, blue, green). Submit each variant as a separate product (for a total of 9 separate products), and submit the same value for the item group ID attribute for each to indicate that these are variants of the same product."

Please, help me to solve this difficult puzzle. I am really struggling.

**Comment by mfhepp** [comment]
Just a bit of historic context: isVariantOf was initially designed so that 
- one could define product properties that are shared / inherited from a base model, 
and then 
- add or replace properties in a deterministic way.

It was initially **not meant to represent rules that define possible product variants,** but instead assuming that product models are materialized (i.e. one entity per each product variant).

The original axioms are documented in here:

http://wiki.goodrelations-vocabulary.org/Axioms

Namely:

http://wiki.goodrelations-vocabulary.org/Axioms#product_variants

I now tend to agree that we likely need a mechanism to specify the range of possible product variants in a rules-based fashion (like "the shirt is available in three colors and six sizes each"), although that creates problems of its own, namely

- reliable identifiers for each variant and
- inventory levels for each variant (any maybe prices and delivery times).

One essential question is whether we want the rules of product variant composition to be 

a) under the command of the publisher (e.g. a server-side Python script that simply materializes all possible variants, or a RESTful API based on a URL pattern), or 
b) whether we define a fully-fledged rule language and provide a vocabulary for representing the respective rules.

Now, as for product model vs. product: There are subtle yet practically relevant differences between product characteristics defined at the level of a product model vs.  those of an actual product.

The original way of representing product characteristics in GoodRelations was organized in a hierarchical fashion, like so

Base Model properties -> Variant -> Actual Product

The exact semantics is specified here:

https://wiki.goodrelations-vocabulary.org/Axioms#product_models

**Bottom-line:** I think it is possible to define a simple, much constrained mechanism so that product variants for commodities like clothing can be represented by rules in markup rather than materialized, but it is slippery ground - because for availability information and ordering, you will need a unique identifier for each variant anyway.

**Outlook:** As search engines are increasingly able to process JSON-LD that is created by active content (e.g. JS) on the client side, **it may be both more feasible and more flexible to develop a best-practice guide on how to generate the schema.org markup for product variants by Javascript components.**

**Comment by jvandriel** [comment]
> _"a simple, much constrained mechanism so that product variants for commodities like clothing"_

Is that enough though, shouldn't this be able to be achieved for non-physical products like SaaS products as well?

Although I do feel that for these it would have been nice if schema.org had chosen for `ProductOrService` as IMO SaaS falls into a gray area in regards to whether it's a service or a product. 

Whichever it is, we wouldn't be able to achieve much these days without SaaS yet it's not a type of product/service which tends to have any globally recognized identifier. At best such companies sometimes sort of 'cheat' in their markup by providing an `sku` value, though 99/100 times its a marketer/SEO that came up with the values (after Google tools warning about them). Most don't actually make use of any sku's for internal systems/purposes.

**Comment by jvandriel** [comment]
@Prlme1, Google has very specific guidelines which describe how they would like you to handle product variants (at this moment in time). I  suggest you [read these](https://developers.google.com/search/docs/specialty/ecommerce/designing-a-url-structure-for-ecommerce-sites#how-google-understands-urls-for-product-variants) and ask any questions you might have about them on [Google's support forum](https://support.google.com/webmasters/community).

**Comment by Tiggerito** [comment]
I re-read the relevant Google docs for Google Search and Google Merchant Centre.

**TL;DR;** I think the closest you can get to comply with both sets of requirements is to mark up a single product with inProductGroupWithID and have an offer per variant that includes the variant SKU or GTIN. Unfortunately, I see no way of describing the variants, and multiple offers may break the rich results, which state the variant should be on its own URL that is canonicalized out, so of no use.

At this time, I only add structured data at the product level. We’ve had several cases where structured data alone can add products to the Merchant Centre.

**Merchant centre:**

https://support.google.com/merchants/answer/6386198

A product page has a single product identified by inProductGroupWithID, which maps to item_group_id in the feed.

ProductModel and ProductGroup do not seem to be mentioned anymore.

https://support.google.com/merchants/answer/7331077

It can have multiple offers that are identified by SKU or GTIN that map to the feeds id. However, I see no way to define what the variant is in the offer.

So offers can be variants.

**On to Google rich snippets…**

https://developers.google.com/search/docs/appearance/structured-data/product#technical-guidelines

Also mentions support for inProductGroupWithID. 

Only mark up one product on a product page. So using things like Product Group will probably break it. I’ve experimented with Product markup on category pages and saw that rich snippets only happened if the category only had one product. 

The document implies only one offer, but does not specifically say so.

It would make sense to allow variants on the product page, where they specify the url for their true location. This could even be just a #green to trigger the page to show variant info.

It references this about giving variants their own URL:

**eCommerce URL recommendations**

https://developers.google.com/search/docs/specialty/ecommerce/designing-a-url-structure-for-ecommerce-sites#how-google-understands-urls-for-product-variants 

It suggests that the canonical tag should be used to have the main product page as the canonical for all the variant pages. This would tell Google not to index the variant pages and not to use their structured data. However, as they would be near duplicates, Google would probably canonicalise them out anyhow. 

So it seems there is no value in marking up variants on variant pages for Google search. Maybe the merchant centre will still check them.

**Comment by jvandriel** [comment]
To respond in particular to what @Tiggerito described about Google's documented desire for canonicalizing product variants to that of a single variant vs real-life situations where web sites offer a product page without having any of the variables set yet (so, variant to be determined still, pretty much like GMC modal for variants identified having to the same [item_group_id]).

It's this where I feel `ProductGroup` (but also `ProductModel` if there is one) might be of use most. e.g.

- https://example.com/product/acme-inc-tshirt/  (clean url = canonical url)
  `ProductGroup` page + variant info all products share, as well as the product variants + their particular properties
- https://example.com/product/acme-inc-t-shirt/?size=xl&color=green 
  `Product` detail page + all it's properties

This way all the variant pages can have their canonical url refer to that of the `ProductGroup` page while making life a whole lot easier for a very big % of the webshops out there because then they don't need to pull all kinds tricks out of their hats to somehow create a system which allows for 1 variant to be the canonical variant. 

Which becomes really painful when such a product is out of stock, because than the canonical has to automatically shift in a cascading manner until there's a variant that is in stock. A seriously stupid and wasteful system which I don't want to know how many billions have been spent on globally already and will be spent more of we don't make it easier for the world by simply letting them provide a `ProductGroup` or `ProductModel` - which describes it all - on the canonical url.

The first example of [my recent post in this thread](https://github.com/schemaorg/schemaorg/issues/1797#issuecomment-1623608970) is actually a result of what @Tiggerito described. That site has a clean canonical url which sort of resembled what [this GMC page](https://support.google.com/merchants/answer/6324507?hl=en) illustrates, but due to the Google Organic Search best practices 'suggestion' for a canonicalized product variant, that website opted to spend dozens and dozens of development hours to accommodate the implementation those 'best practices. 

Yet they already had the perfect system in place - nearly out of the box - to be able to generate a page of which the `ProductGroup` would have been the `mainEntityOfPage` (and no, that particular shop also has no `ProductModel` info, like most shops don't).

@Tiggerito unfortunately still is very correct in saying:  _"So it seems there is no value in marking up variants on variant pages for Google search."_

While also being right in saying: _"So using things like Product Group will probably break it"_. I've extensively tested this over the last couple of months and so far the outcome of `ProductGroup` + all `Product` variants on a page = NO Rich Results. And since actually negatively impacts revenue of shops they won't publish `ProductGroup` markup unless there at least some sign of (coming) support.

At this moment the only sort of practical use case I've been able to find are SaaS subscription overview pages, which due to their nature already aren't eligible for Google's Rich Results. But beyond that I can find no candidates that are willing to take a chance with publishing `ProductGroup`  markup.

**Comment by Prlme1** [comment]
I went to the nike shop and read off the schema they have. This is how the major shops do. they put everything in "AggragateOffer" section. However, they ignore the sizes for some reason. and I still cannot understand how this appears in the Google Merchant Center, since all variants have to be a separate product:

</script>
<div id="nikeID-modal-container" class="nike-id-builder-container" hidden=""></div>
<div id="modal-root"></div>
<script type="application/ld+json">
{
  "@context": "https://schema.org",
  "@type": "Product",
  "name": "Nike Spark Women's Shoes",
  "image": "https://static.nike.com/a/images/t_default/b438f82a-62a3-49a8-9ded-fd1d281761a9/spark-shoes-zPVQHZ.png",
  "description": "Find the Nike Spark at Nike.com.",
  "brand": {
    "@type": "Brand",
    "name": "Nike",
    "logo": "https://www.nike.com/Nike_Swoosh_Logo_Black_original.jpg"
  },
  "offers": {
    "@type": "AggregateOffer",
    "lowPrice": "8957",
    "highPrice": "12795",
    "priceCurrency": "INR",
    "availability": "https://schema.org/InStock",
    "offerCount": 5,
    "offers": [
      {
        "@type": "Offer",
        "url": "https://www.nike.com/in/t/spark-shoes-zPVQHZ",
        "price": 8957,
        "priceCurrency": "INR",
        "itemOffered": {
          "@type": "Product",
          "name": "Nike Spark Women's Shoes",
          "model": "14084002",
          "releaseDate": "2022-10-06T16:00:00.000Z",
          "color": "Barely Rose/Pink Oxford/Light Soft Pink/Dark Smoke Grey"
        },
        "seller": {
          "@type": "Organization",
          "name": "Nike"
        },
        "availability": "https://schema.org/OutOfStock"
      },
      {
        "@type": "Offer",
        "url": "https://www.nike.com/in/t/spark-shoes-zPVQHZ",
        "price": 8957,
        "priceCurrency": "INR",
        "itemOffered": {
          "@type": "Product",
          "name": "Nike Spark Women's Shoes",
          "model": "1003839125",
          "releaseDate": "2023-01-30T16:00:00.000Z",
          "color": "Summit White/Rush Fuchsia/Ocean Bliss/Metallic Silver"
        },
        "seller": {
          "@type": "Organization",
          "name": "Nike"
        },
        "availability": "https://schema.org/InStock"
      },
      {
        "@type": "Offer",
        "url": "https://www.nike.com/in/t/spark-shoes-zPVQHZ",
        "price": 12795,
        "priceCurrency": "INR",
        "itemOffered": {
          "@type": "Product",
          "name": "Nike Spark Women's Shoes",
          "model": "14018627",
          "releaseDate": "2022-11-22T16:00:00.000Z",
          "color": "Phantom/Sand Drift/White/Dark Smoke Grey"
        },
        "seller": {
          "@type": "Organization",
          "name": "Nike"
        },
        "availability": "https://schema.org/InStock"
      },
      {
        "@type": "Offer",
        "url": "https://www.nike.com/in/t/spark-shoes-zPVQHZ",
        "price": 12795,
        "priceCurrency": "INR",
        "itemOffered": {
          "@type": "Product",
          "name": "Nike Spark Women's Shoes",
          "model": "14141745",
          "releaseDate": "2022-10-06T16:00:00.000Z",
          "color": "Black/Pure Platinum/White/Metallic Silver"
        },
        "seller": {
          "@type": "Organization",
          "name": "Nike"
        },
        "availability": "https://schema.org/InStock"
      },
      {
        "@type": "Offer",
        "url": "https://www.nike.com/in/t/spark-shoes-zPVQHZ",
        "price": 12795,
        "priceCurrency": "INR",
        "itemOffered": {
          "@type": "Product",
          "name": "Nike Spark Women's Shoes",
          "model": "1010134718",
          "releaseDate": "2023-03-26T16:00:00.000Z",
          "color": "White/Game Royal/Picante Red/Black"
        },
        "seller": {
          "@type": "Organization",
          "name": "Nike"
        },
        "availability": "https://schema.org/InStock"
      }
    ]
  },
  "aggregateRating": {
    "@type": "AggregateRating",
    "ratingValue": "4.9",
    "reviewCount": 9,
    "bestRating": "5",
    "worstRating": "1"
  }
}
</script>

**Comment by Tiggerito** [comment]
@Prlme1 using AggregateOffer means the page is not eligible for the merchant centre:

https://developers.google.com/search/docs/appearance/structured-data/product#aggregate-offer-properties

They assume that means you are an affiliate.

Could you provide an example URL where your example structured data is causing rich snippets. They have a strange nested offer structure which might end up being acceptable.

**Comment by Prlme1** [comment]
Thank you for sharing this information! Regarding your question about Nike shoes on Google, the products are displayed as goods in the image tab. However, it's uncertain how comprehensive these rich results are.

**Comment by briansackett** [comment]
Hello, It is great to see that this issue is being addressed after having made an educated guess for aggregating products with duplicate content back in 2019. I have spent countless hours trying to figure out how this simple, and very common circumstance should be treated in the eyes of google. If anyone can help in lehman terms that would be much appreciated. It is quite difficult to drill down to a source of truth.

For simplification, i will separate variation tiers by (A, B, C) and will address each tier as defined below.

-Product Category ( A )
  - Product Line (AA) | (AB)
    - Product (AAA) (AAB) (AAC) | (ABA) (ABB) (ABC)

**Topics of Confusion**

- Where should the PRODUCT structured data exist? Should the product category structured data drill down to the product? Should it be on every page the product is listed?(A, AA, AAA) Is listing it multiple times a bad thing?
 
- What is the the best way of aggregating structured data on the "Product Line" page? 
            1. Product->Offers->AggregateOffers->Product[] ** this will error as the product line itself has no offer which is a required field.
            2. ItemList->itemListElement->Product[] **this is what we are doing and google seems to not like this
            3. ProductGroup. **if this is the best choice, why is hasVariant a 1:1 data relation? what is the proper syntax for this new method? 
- Proper Listing/Aggregating Syntax for Google
          1. Will the inverse Property on Product for aggregation (isVariantOf) register Variant to Parent if it is done on a different page (ie. ProductGroup defined on Product Line page, and reference via @id property on Product Page)?   
          2. Should I be listing Products in separate JSON objects or nested within an ItemList property of the Aggregate model?

-   What page is the most important in terms of Content/Duplicate Content. This is not directly related, but is very similar in terms of what google is trying to achieve. The Product Category, Product Line and Product Pages are all relevant in terms of product drill down, but they all have nearly identical information.  When a customer knows what they are looking for, they should be able to find it (ie. search via SKU to a Product specific page), but at what cost to the vendors Page Rank score? The reason I bring this up is because this "Pillar Page" of information would seemingly be the place where you would want to put all of your structured data and content in hopes of indexing and ranking highly with many product snippets associated.


Thank you for your help, the more i read the more contradictions i find from different parts of the thread, or different information online.. hoping to get something firm to stand on.

**Comment by Alireza-Naji** [comment]
hello dear @jvandriel . i would be grateful if you would take the time to review my suggestion about "varianttype" and if you provide me with your feedback. thank you for your time and consideration.

**Rationale for "VariantType"** 

after conducting a thorough analysis of the entire conversation and the provided texts, the introduction of the VariantType stems from a need for a more robust and standardized way to represent variable features within a group of products. The primary motivations for this proposal are:

Semantic Enhancement:
The current data lacks a standardized way to represent and categorize variable features of products.
VariantType introduces a structured approach to add semantic meaning to variable features, enhancing the overall understanding of the products.

User Experience Improvement:
With products having variable features such as different colors, sizes, or materials, there is a need to provide users with a more intuitive and efficient means of navigating through these variations.
VariantType facilitates an improved user experience by categorizing and presenting products based on their variable features.

Data Modeling Flexibility:
as products evolve, new variable features may be introduced. VariantType is designed to be extendable, allowing for the incorporation of additional information and features in a flexible manner.
this flexibility ensures that the proposed solution is adaptable to future changes in product attributes.

Categorization and Recommendation:
The introduction of VariantType allows for effective categorization of products based on variable features, providing a foundation for developing recommendation systems.
Users can benefit from a more organized presentation of products, making it easier for them to find and choose the products that meet their specific requirements.

Consistency Across Products:
a standardized approach like VariantType ensures consistency in representing variable features across different products, making it easier for developers, applications, and users to understand and work with the data.

By addressing these considerations, the VariantType proposal aims to provide a solution that not only fills the current gap in data representation but also anticipates future needs and changes in the landscape of product information. The objective is to enhance both the backend structuring of data and the frontend user experience, creating a more comprehensive and user-friendly system for handling variable product features.

**Proposal Explanation:**

Using VariantType:
Creating a variable named VariantType that can serve as a tag for variable characteristics in related products. This VariantType can act as a criterion for categorizing and recommending associated products. For example, for products available in different colors, VariantType could be "Color."

This proposal allows you to enhance semantic structure, improve product information based on variable features, and assist users in quickly finding the products they need.

First Example:

VariantType is a variable used to model types of variable features within a group of products. This variable adds meaning to product information and helps systems and users quickly understand the differences and various features of products.

VariantType Features:

   Name:
   Determine a name for each VariantType, such as "Color," "Size," or "Material."

   Variant Values:
   Define different variable values for each VariantType. For example, for a VariantType named "Color," variable values could be "Red," "Blue," and "Green."

   Associated Products:
   Each VariantType can reference associated products that have variable features.

   Categorization:
   The ability to categorize products based on VariantType and their variable values.

   Extendable Information:
   Add extendable information to each VariantType to model various features that may be added in the future.

Example:
For a product available in different colors:

   VariantType: Color
       Variable Values:
           Red
           Blue
           Green

Using VariantType allows for easy and improved modeling of variable product information, enabling users to quickly find the products they need.

Second Example:

VariantType is a variable used to model types of variable features within a group of products. This variable adds meaning to product information and helps systems and users quickly understand the differences and various features of products.

**VariantType Features:**

   Name:
   Determine a name for each VariantType, such as "Color," "Size," or "Material."

   Variant Values:
   Define different variable values for each VariantType. For example, for a VariantType named "Size," variable values could be "Small," "Medium," and "Large."

   Associated Products:
   Each VariantType can reference associated products that have variable features.

   Categorization:
    The ability to categorize products based on VariantType and their variable values.

    Extendable Information:
    Add extendable information to each VariantType to model various features that may be added in the future.

example:
for a product available in different sizes:

    VariantType: Size
        Variable Values:
            Small
            Medium
            Large

using VariantType allows for easy and improved modeling of variable product information, enabling users to quickly find the products they need.

```
{
  "@context": "https://schema.org",
  "@type": "VariantType",
  "name": "Color",
  "variantValues": [
    {
      "@type": "PropertyValue",
      "name": "Red"
    },
    {
      "@type": "PropertyValue",
      "name": "Blue"
    },
    {
      "@type": "PropertyValue",
      "name": "Green"
    }
  ],
  "associatedProducts": [
    {
      "@type": "Product",
      "name": "Product A",
      "color": "Red"
    },
    {
      "@type": "Product",
      "name": "Product B",
      "color": "Blue"
    },
    {
      "@type": "Product",
      "name": "Product C",
      "color": "Green"
    }
  ]
}
```

this code illustrates the structure of VariantType, its features, and an example for a product with variable colors.

thank you

**Comment by jvandriel** [comment]
@Alireza-Naji, could you please open a new issue for your request? You're asking for something that is outside the scope of this thread.

**Comment by jvandriel** [comment]
@briansackett, you seem to have all kinds of questions about how you should do things for Google. I suggest you take your questions to [their forum](https://support.google.com/webmasters/threads?hl=en&thread_filter=(category:structured_data)&sjid=8486899563522657978-EU). We do not handle that here.

**Comment by alex-jansen** [comment]
Closing since variant support was added in Release 9.0 per the release notes.
